pkg/oc/cli/admin/mustgather/mustgather.go

```diff
@@ -0,0 +1,367 @@
+	cmd.Flags().StringVar(&o.DestDir, "dest-dir", "", "Set a specific directory on the local machine to write gathered data to.")
+	cmd.Flags().BoolVar(&o.Keep, "keep", false, "Do not delete temporary resources when command completes.")
+	_ = cmd.Flags().MarkHidden("keep")
+
+	return cmd
+}
+
+func (o *Options) Complete(f kcmdutil.Factory, cmd *cobra.Command, args []string) error {
+	if i := cmd.ArgsLenAtDash(); i != -1 && i < len(args) {
+		o.Command = args[i:]
+	} else {
+		o.Command = args
+	}
+
+	var err error
+
+	if o.ClientConfig, err = f.ToRESTConfig(); err != nil {
+		return err
+	}
+
+	if o.KubeClient, err = kubernetes.NewForConfig(o.ClientConfig); err != nil {
+		return err
+	}
+
+	if len(o.DestDir) == 0 {
+		o.DestDir = fmt.Sprintf("must-gather.local.%06d", rand.Int63())
+	}
+
+	if len(o.Image) == 0 {
+		// TODO lookup cluster specific default
+		o.Image = "quay.io/openshift/origin-must-gather:v4.0"
+	}
+
+	return nil
+}
+
+type Options struct {
+	genericclioptions.IOStreams
+
+	ClientConfig *rest.Config
+	KubeClient   kubernetes.Interface
+
+	NodeName string
+	DestDir  string
+	Image    string
+	Command  []string
+	Keep     bool
+}
+
+// Run creates and runs a must-gather pod.d
+func (o *Options) Run() error {
+	if len(o.ClientConfig.BearerToken) != 0 {
+		return fmt.Errorf("cannot run must-gather with a token based kubeconfig")
+	}
+	if len(o.Image) == 0 {
+		return fmt.Errorf("missing an image")
+	}
+
+	// TODO This command will:
+	// [x] Create a temporary namespace
+	// [x] Create a secret an upload the local kubeconfig content into it
+	// [ ] Look up the image to use for gathering information
+	// [x] Create a pod that runs the 'gather' command in that image
+	// [x] Wait for completion of that command
+	// [xs] Copy the gathered data locally
+
+	var err error
+
+	// create namespace
+	ns, err := o.KubeClient.CoreV1().Namespaces().Create(&corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "openshift-must-gather-",
+			Labels: map[string]string{
+				"openshift.io/run-level": "0",
+				"app": "must-gather",
+			},
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if !o.Keep {
+		defer func() {
+			err = o.KubeClient.CoreV1().Namespaces().Delete(ns.Name, nil)
+		}()
+	}
+
+	clusterRoleBinding, err := o.KubeClient.RbacV1().ClusterRoleBindings().Create(o.newClusterRoleBinding(ns.Name))
+	if err != nil {
+		return err
+	}
+	if !o.Keep {
+		defer func() {
+			err = o.KubeClient.RbacV1().ClusterRoleBindings().Delete(clusterRoleBinding.Name, &metav1.DeleteOptions{})
+		}()
+	}
+
+	// create pod
+	pod, err := o.KubeClient.CoreV1().Pods(ns.Name).Create(o.newPod(o.NodeName))
+	if err != nil {
+		return err
+	}
+
+	// wait for pod to be running (i.e. gather has completed)
+	podPhase, err := o.waitForPodNotPending(pod)
+	if err != nil {
+		return err
+	}
+	if podPhase != corev1.PodRunning {
+		return fmt.Errorf("pod is not running: %v", podPhase)
+	}
+
```

Comment:
Add a `// TODO, confirm that must-gather has finish`

Reply:
must-gather is run in an init container. Pod is `Pending` while it's running, `Running` if successful, and something else if something went wrong.